rasa/nlu/classifiers/keyword_intent_classifier.py

```diff
@@ -1,32 +1,156 @@
-from typing import Any, Optional, Text
+import os
+import logging
+import typing
+import re
+from typing import Any, Dict, Optional, Text
 
+from rasa.nlu import utils
 from rasa.nlu.components import Component
 from rasa.nlu.training_data import Message
 
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    from rasa.nlu.config import RasaNLUModelConfig
+    from rasa.nlu.training_data import TrainingData
+    from rasa.nlu.model import Metadata
+    from rasa.nlu.training_data import Message
+
 
 class KeywordIntentClassifier(Component):
+    """Intent classifier using simple keyword matching.
+
+
+    The classifier takes a list of keywords and associated intents as an input.
+    A input sentence is checked for the keywords and the intent is returned.
+
+    """
 
     provides = ["intent"]
 
-    his = ["hello", "hi", "hey"]
+    defaults = {"case_sensitive": True}
+
+    def __init__(
+        self,
+        component_config: Optional[Dict[Text, Any]] = None,
+        intent_keyword_map: Optional[Dict] = None,
+    ):
+
+        super(KeywordIntentClassifier, self).__init__(component_config)
 
-    byes = ["bye", "goodbye"]
+        self.intent_keyword_map = intent_keyword_map or {}
+
+        self.re_case_flag = (
+            0 if self.component_config["case_sensitive"] else re.IGNORECASE
+        )
+
+    def train(
+        self,
+        training_data: "TrainingData",
+        cfg: Optional["RasaNLUModelConfig"] = None,
+        **kwargs: Any
+    ) -> None:
+
+        duplicate_examples = set()
+        for ex in training_data.training_examples:
+            if (
+                ex.text in self.intent_keyword_map.keys()
+                and ex.get("intent") != self.intent_keyword_map[ex.text]
+            ):
+                duplicate_examples.add(ex.text)
+                logger.warning(
+                    "Keyword '{}' is an example of intent '{}' and of "
+                    "intent '{}', it will be removed from the list of "
+                    "keyword.\n"
+                    "Remove (one of) the duplicates from the training data."
+                    "".format(
+                        ex.text, self.intent_keyword_map[ex.text], ex.get("intent")
+                    )
+                )
+            else:
+                self.intent_keyword_map[ex.text] = ex.get("intent")
+        for keyword in duplicate_examples:
+            self.intent_keyword_map.pop(keyword)
+            logger.debug(
+                "Removed '{}' from the list of keywords because it was "
+                "a keyword for more than one intent.".format(keyword)
+            )
+
+        self._validate_keyword_map()
+
+    def _validate_keyword_map(self):
+        ambiguous_mappings = []
+        for ex1, intent1 in self.intent_keyword_map.items():
```

Comment:
```suggestion
        for text1, intent1 in self.intent_keyword_map.items():
```

Reply:
`text` is always used for the text entered by the user, i also don't like `ex1` a lot but i think it's better to keep these names distinct

Reply:
but that is not an example, what about `key_text`?